src/mscorlib/shared/System/IO/FileStream.cs

```diff
@@ -81,8 +81,31 @@ public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferS
 
         [Obsolete("This constructor has been deprecated.  Please use new FileStream(SafeFileHandle handle, FileAccess access, int bufferSize, bool isAsync) instead, and optionally make a new SafeFileHandle with ownsHandle=false if needed.  http://go.microsoft.com/fwlink/?linkid=14202")]
         public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferSize, bool isAsync)
-            : this(new SafeFileHandle(handle, ownsHandle), access, bufferSize, isAsync)
         {
+            SafeFileHandle safeHandle = new SafeFileHandle(handle, ownsHandle: ownsHandle);
+            try
+            {
+                ValidateAndInitFromHandle(safeHandle, access, bufferSize, isAsync);
+            }
+            catch
+            {
+                // We don't want to take ownership of closing passed in handles
+                // *unless* the constructor completes successfully.
+                GC.SuppressFinalize(safeHandle);
+
+                // This would also prevent Close from being called, but is unnecessary
+                // as we've removed the object from the finalizer queue.
+                //
+                // safeHandle.SetHandleAsInvalid();
+                throw;
+            }
+
+            _access = access;
+            _useAsyncIO = isAsync;
+
+            // As the handle was passed in, we must set the handle field at the very end to
```

Comment:
I do not understand why you are doing this. I think it would be simpler and easier to understand if this is implemented like this:

```
public FileStream(IntPtr handle, FileAccess access, bool ownsHandle, int bufferSize, bool isAsync) 
{
  try 
  { 
       ValidateAndInitFromHandle(safeHandle, access, bufferSize, isAsync); 
   }
   catch
   {
       // We don't want to take ownership of closing passed in handles 
       // *unless* the constructor completes successfully. 
       GC.SuppressFinalize(safeHandle);
       _fileHandle = null;
   }
}
```